[popover] Check for modal flag instead of open attribute

rwlbuis opened the pull request:
#### 62bb3c714cc988dad1913f7c2b6193182b7f4318
<pre>
[popover] Check for modal flag instead of open attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=257412">https://bugs.webkit.org/show_bug.cgi?id=257412</a>

Reviewed by Tim Nguyen.

Check for modal flag instead of open attribute in checkPopoverValidity check:
<a href="https://github.com/whatwg/html/pull/9344">https://github.com/whatwg/html/pull/9344</a>

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-top-layer-combinations.html:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::show):
(WebCore::HTMLDialogElement::close):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::checkPopoverValidity):

Canonical link: <a href="https://commits.webkit.org/264876@main">https://commits.webkit.org/264876@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fef69f8ea541d78cb36a50d48cd0eb7a9103f0be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8927 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9215 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9433 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10580 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8907 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8935 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11202 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9182 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11755 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9073 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10062 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10739 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7357 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8162 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15650 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8466 "4 api tests failed or timed out") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8310 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11638 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8804 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7184 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8059 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/8069 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2169 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12271 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8551 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1456,7 +1456,8 @@ ExceptionOr<void> HTMLElement::hidePopoverInternal(FocusPreviousElement focusPre
 
-    removeFromTopLayer();
+    if (isInTopLayer())
+        removeFromTopLayer();
```
Is there a particular reason we need this change? (and would the spec need changing as well?)

rwlbuis replied:
I saw the ASSERT in removeFromTopLayer happen when hiding the popover on cleanup/node removal. I think this can happen when dialog and popover API are intermixed. Probably the popover+dialog is shown, the popover+dialog closed and then removed. Note thay HTMLDialogElement::close tests for isInTopLayer.

nt1m replied:
I think I see the issue here, `dialog.close()` is also removing the popover from the top layer. Not sure this is intended. Perhaps we should only remove the popover from the top layer if the "is modal" flag is set to true in `HTMLDialogElement::close()`. Will suggest a spec change.

nt1m replied:
@rwlbuis Can you check if the latest spec change: https://github.com/whatwg/html/pull/9344 resolves the issue without the need for changing `hidePopover()` ?

rwlbuis replied:
It seems to work in my release build, but I did not try debug yet. Anyway hopefully mac-AS-debug-wk2 will show whether the new patch asserts/works or not.

rwlbuis committed:
[popover] Check for modal flag instead of open attribute
https://bugs.webkit.org/show_bug.cgi?id=257412

Reviewed by Tim Nguyen.

Check for modal flag instead of open attribute in checkPopoverValidity check:
https://github.com/whatwg/html/pull/9344

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-top-layer-combinations.html:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::show):
(WebCore::HTMLDialogElement::close):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::checkPopoverValidity):

Canonical link: https://commits.webkit.org/264876@main
